### PR TITLE
feat: add front menu

### PR DESCRIPTION
**Pull request by Jocs**
| Q                | A
| ---------------- | ---
| Bug fix?         | no
| New feature?     | yes
| BC breaks?       | no
| Deprecations?    | no
| New tests added? | not needed
| License          | MIT

### Description

[Description of the bug or feature]

**Features**

- [X] Add front menu.
- [x] add short cut of `duplicate`, `delete`, `create paragraph`.

**Bugs**

- [x] bug when update `blockquote` to `paragraph`.
- [x] A paragraph turn into `paragraph`, and the text in this paragraph disappeared.
- [x] Front matter can not be duplicated. One document only has one `front matter`.

Add a front menu by click the front menu icon.
屏幕快照 2019-04-06 上午1 47 12

--

### CODE REVIEW

**Comment by Jocs** [comment]
@fxha Can we add this feature to `v0.14.0`?

File: src/main/shortcutHandler.js [hunk 1]
```diff
@@ -41,6 +41,9 @@ class Keybindings {
       ['editCopyAsMarkdown', 'CmdOrCtrl+Shift+C'],
       ['editCopyAsPlaintext', 'CmdOrCtrl+Shift+V'],
       ['editSelectAll', 'CmdOrCtrl+A'],
+      ['editDuplicate', 'Shift+CmdOrCtrl+P'],
+      ['editCreateParagraph', 'Shift+CmdOrCtrl+N'],
+      ['editDeleteParagraph', 'Shift+CmdOrCtrl+D'],
```
Comment (fxha) [inline src/main/shortcutHandler.js]: Please add the shortcuts to the [documentation](https://github.com/marktext/marktext/blob/master/doc/KEYBINDINGS.md).

**Comment by Jocs** [comment]
>Paragraph to table, math block, HTML and maybe more doesn't work

Do you mean it create a `table`, `math`, `HTML` after the current paragraph?

**Comment by Jocs** [comment]
>Issue with task list item because it's detected as bullet list when coming from source code mode (see #870 (comment))

Yes, I think we need to treat `task list` as a separate list, not a bullet list. only in this way, we can fix a lot problem. I'll fix this in 0.15.0.

**Comment by Jocs** [comment]
@fxha fixed `Bullet to ordered list issue`.

**Comment by Jocs** [comment]
>I think we could add this to v0.14.0 and block the release longer. Maybe add it to v0.14.1 that fixes not such important bugs?

I think it depends on how many bugs this will feature bring about. if there is no serious bugs, I think we can release the `pre release` version after this PR merged.

File: src/muya/lib/parser/marked/lexer.js [hunk 2]
```diff
@@ -207,11 +207,11 @@ Lexer.prototype.token = function (src, top) {
       src = src.substring(cap[0].length)
       bull = cap[2]
       let isOrdered = bull.length > 1 && /\d{1,9}/.test(bull)
-
+      console.log(bull.length, cap)
       this.tokens.push({
         type: 'list_start',
         ordered: isOrdered,
-        listType: bull.length > 1 ? (/\d{1,9}/.test(bull) ? 'order' : 'task') : 'bullet',
+        listType: bull.length > 1 ? 'order' : (/^( {0,3})([-*+]) \[([xX]{1}| )\]/.test(cap[0]) ? 'task' : 'bullet'),
```
Comment (fxha) [inline src/muya/lib/parser/marked/lexer.js]: Why not just `/^( {0,3})([-*+]) \[([ xX])\]/`?

File: src/muya/lib/parser/marked/lexer.js [hunk 3]
```diff
@@ -257,7 +258,7 @@ Lexer.prototype.token = function (src, top) {
           this.tokens.push({
             type: 'list_start',
             ordered: isOrdered,
-            listType: bull.length > 1 ? (/\d{1,9}/.test(bull) ? 'order' : 'task') : 'bullet',
+            listType: bull.length > 1 ? 'order' : (/^( {0,3})([-*+]) \[([xX]{1}| )\]/.test(itemWithBullet) ? 'task' : 'bullet'),
```
Comment (fxha) [inline src/muya/lib/parser/marked/lexer.js]: Why not just `/^( {0,3})([-*+]) \[([ xX])\]/`?

File: src/muya/lib/parser/marked/lexer.js (hunk 2 above)
Comment (fxha) [reply src/muya/lib/parser/marked/lexer.js]: Task list items are not clearly defined. According to GFM a task list item can be `- [ ]` but GitHub renders a task list item only if `- [ ] a`. Maybe we should check for `/^( {0,3})([-*+]) \[([ xX])\] /` like all popular parsers?

**Comment by Jocs** [comment]
>Paragraph Turn Into table, math block etc still not working

Do you mean it create `table`, `math` block bellow?

File: src/muya/lib/parser/marked/lexer.js (hunk 2 above)
Comment (Jocs) [reply src/muya/lib/parser/marked/lexer.js]: What's the meaning `just`?

**Comment by fxha** [comment]
> Do you mean it create table, math block bellow?

Yeah

File: src/muya/lib/parser/marked/lexer.js [hunk 4]
```diff
@@ -207,11 +207,11 @@ Lexer.prototype.token = function (src, top) {
       src = src.substring(cap[0].length)
       bull = cap[2]
       let isOrdered = bull.length > 1 && /\d{1,9}/.test(bull)
```
Comment (fxha) [inline src/muya/lib/parser/marked/lexer.js]: BTW you can also change this to `bull.length > 1`.

File: src/muya/lib/parser/marked/lexer.js (hunk 2 above)
Comment (Jocs) [reply src/muya/lib/parser/marked/lexer.js]: My fault!

**Comment by Jocs** [comment]
Now, paragraph => any is wrong when any is on of `front matter`, `horizontal line`, `table`, `math`, `html`.

I'll disable paragraph turn to `front matter` and horizontal line. because it's nonsense behavior.

I think paragraph turn into `html`, `math` is OK, I'll modify it later. But what do think of turn into table? put all the paragraph text in the first cell?

**Comment by fxha** [comment]
> But what do think of turn into table? put all the paragraph text in the first cell?

I don't really know, put the paragraph in the first cell sound good.

**Comment by Jocs** [comment]
>I don't really know, put the paragraph in the first cell sound good.

I have done this PR, I also disabled paragraph turn into table, because I found `Notion(a note taking app)` also disable it too.

**Comment by Jocs** [comment]
Updated!

**Comment by Jocs** [comment]
>Multiselection doesn't work too:

Yes, if selection in two paragraphs, front menu should be disabled.

**Comment by Jocs** [comment]
>Multiselection doesn't work too:

Fixed, if the selection in two paragraphs, disable the front menu.
if the selection is in two lines, disable paragraph turn into heading.

**Pull request merged**